src/check_python_versions/sources/tox.py

```diff
@@ -99,13 +100,19 @@ def tox_env_to_py_version(env: str) -> Optional[Version]:
     If the environment name has dashes, only the first part is considered,
     e.g. py34-django20 becomes '3.4', and jython-docs becomes 'jython'.
     """
+    global has_dotted_env
+    has_dotted_env = False  # reset global state before check
```

Comment:
Can we avoid global state _please_?

IIIRC I had a function to parse tox envlist names and convert them to Version objects.  It should handle optional dots.

The update functions should instead check whether the envlists were using dots or not, just like they currently check whether the envlist uses spaces after commas.

Reply:
the main way I followed when developing was to choose the format to be used in the whole file:
1. if no env has dot, the update should put envs without dot
2. if envs have at least one dot, the update should put envs with dot
3. in mixed case env, it's forced to "change all env with dot"

I know that point 3 is little risky, because developer should update the environments in the remaining part of the file.
I've no idea how to keep "mixed style env" and I don't know what's your point.

Reply:
Sure, I agree with points (1) and (2) and don't care about point (3) -- normalizing to one style is fine, I don't particularly care which style it should be.

My point was that this doesn't need a global switch, a local check inside the update_...() function that then passes the detected style to any helpers should suffice.

Reply:
perhaps I didn't express correctly the drawback about third point : in case of this starting state

```
[tox]
envlist = py3.8, py3.9, py310, py311
...
[testenv]
setenv =
    {py3.8, py3.9, py310, py311}: <<do something here>>
commands =
    py3.8:  <<do something here>>
    py3.9:  <<do something here>>
    py310:  <<do something here>>
    py311:  <<do something here>>
```

if the dot isn't stored in some object that identifies what's found in envlist (Version object are created), it can happen that after the update the state becomes (e.g. deciding to force dot removal way)
```
[tox]
envlist = py38, py39, py310, py311
...
[testenv]
setenv =
    {py3.8, py3.9, py310, py311}: <<do something here>>    <<< HERE NO CHANGES, misalignment with testenv
commands =
    py3.8:  <<do something here>>    <<< HERE NO CHANGES, misalignment with testenv
    py3.9:  <<do something here>>    <<< HERE NO CHANGES, misalignment with testenv
    py310:  <<do something here>>
    py311:  <<do something here>>
```

as you can see, the envlist is updated removing the dots on 3.8 and 3.9, but the remaining part of the file was not touched , and this led to issue if no review is done AFTER check-python-version was run.

Instead, keeping the 'dot' information as I did, the already existing version into envlist are not changed, so everything still match in testenv section.

I hope that his had clarified the reason about my recent changes.

Reply:
I see.  I would hope that tox is smart enough to consider e.g. `py38` and `py3.8` factors as equivalent, so the configuration would continue to work despite the misalignment.

My view on this remains: preserving a preexisting mixture of styles is extra effort, I don't mind if it's done, but it's not a requirement for the PR to be merged. 

Adding mutable global state or a thematically-unrelated attribute to a generic Version class are deal-breakers.

If I wanted to preserve a mixture of styles (which I don't want to do), I'd create a supplementary mapping of Version -> str in update_tox_envlist:

```python
old_spelling = {}
for env in parse_envlist(envlist):
    ver = tox_env_to_py_version(env)
    if ver:
        old_spelling[ver] = env
```

and then reuse it when computing new_envs:

```python
new_envs = [
    old_spelling.get(ver, toxenv_for_version(ver))
    for ver in new_versions
]
```

This way the state is local and Version doesn't need to be extended with complications that affect equality comparisons.

Reply:
I'm more worried about the `# Try to preserve braced groups` branch.  Does that continue to work?  Does it support dotted versions?  E.g.  will

```dosini
[tox]
envlist = py{3.9,3.10}{,-foo}
```

be converted to

```dosini
[tox]
envlist = py{3.9,3.10,3.11}{,-foo}
```

after a `check-python-versions --add 3.11`?

Reply:
> I see. I would hope that tox is smart enough to consider e.g. `py38` and `py3.8` factors as equivalent, so the configuration would continue to work despite the misalignment.


I've just done some tests, seems not so smart as you hope:

```
[tox:tox]
envlist = py3.10

[testenv]
commands =
    py3.10: echo dotted
    py310: echo not-dotted
    echo after
```

if I run `poetry run tox`, the message I see on screen are `dotted` and `after` . Switching to `envlist = py310` I see on screen are `not-dotted` and `after` .